presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -0,0 +1,440 @@
+    private final Map<String, TypeVariableConstraint> typeVariableConstraints;
+
+    public SignatureBinder(TypeManager typeManager, Signature declaredSignature, boolean allowCoercion)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.declaredSignature = requireNonNull(declaredSignature, "parametrizedSignature is null");
+        this.allowCoercion = allowCoercion;
+        this.typeVariableConstraints = declaredSignature.getTypeVariableConstraints()
+                .stream()
+                .collect(toMap(TypeVariableConstraint::getName, t -> t));
+    }
+
+    public Optional<Signature> matchAndBindSignature(List<? extends Type> actualArgumentTypes)
+    {
+        Optional<BoundVariables> boundVariables = matchAndBindSignatureVariables(actualArgumentTypes);
+        if (!boundVariables.isPresent()) {
+            return Optional.empty();
+        }
+        return Optional.of(bindVariables(declaredSignature, boundVariables.get(), actualArgumentTypes.size()));
+    }
+
+    public Optional<BoundVariables> matchAndBindSignatureVariables(List<? extends Type> actualArgumentTypes)
+    {
+        List<TypeSignature> expectedArgumentSignatures = declaredSignature.getArgumentTypes();
+        boolean variableArity = declaredSignature.isVariableArity();
+        BoundVariables.Builder boundVariablesBuilder = BoundVariables.builder();
+        if (!matchArguments(expectedArgumentSignatures, actualArgumentTypes, variableArity, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+        BoundVariables boundVariables = boundVariablesBuilder.build();
+        checkAllTypeVariablesAreBound(boundVariables);
+        return Optional.of(boundVariables);
+    }
+
+    public Optional<BoundVariables> matchAndBindSignatureVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)
+    {
+        BoundVariables.Builder boundVariablesBuilder = BoundVariables.builder();
+
+        TypeSignature expectedReturnTypeSignature = declaredSignature.getReturnType();
+        if (!matchAndBindArgument(expectedReturnTypeSignature, actualReturnType, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+
+        List<TypeSignature> expectedArgumentSignatures = declaredSignature.getArgumentTypes();
+        boolean variableArity = declaredSignature.isVariableArity();
+        if (!matchArguments(expectedArgumentSignatures, actualArgumentTypes, variableArity, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+
+        BoundVariables boundVariables = boundVariablesBuilder.build();
+        checkAllTypeVariablesAreBound(boundVariables);
+        return Optional.of(boundVariables);
+    }
+
+    private boolean matchArguments(
+            List<TypeSignature> expectedArgumentSignatures,
```

Comment:
I'd call these `expectedTypes` and `actualTypes`. They are shorter, easier to read and don't sacrifice on clarity.